src/bit_str.jl

```diff
@@ -276,210 +207,78 @@ function Base.iterate(bit::BitStr, state::Integer=1)
-
-```jldoctest
-julia> setbit(bit"1011", 0b1100)
-1111 (15)
-
-julia> setbit(bit"1011", 0b0100)
-1111 (15)
-
-julia> setbit(bit"1011", 0b0000)
-1011 (11)
-```
+    bint_r(b::BitStr) -> Integer
 """
-setbit(b::BitStr{T}, mask::Integer) where T = setbit(b.val, T(mask)) |> bit(len=length(b))
-
+bint_r(b::BitStr{N}) where N = Integer(breflect(b))
 
 """
-    anyone(b::BitStr, mask::Integer) -> Bool
-
-Return `true` if any masked position of index is 1.
-
-# Example
-
-`true` if any masked positions is 1.
-
-```jldoctest
-julia> anyone(bit"1011", 0b1001)
-true
-
-julia> anyone(bit"1011", 0b1100)
-true
-
-julia> anyone(bit"1011", 0b0100)
-false
-```
+    bint(b::BitStr) -> Integer
 """
-anyone(b::BitStr{T}, mask::Integer) where T = anyone(b.val, mask) |> bit(len=length(b))
+bint(b::BitStr) = Integer(b)
 
 """
-    allone(b::BitStr, mask::Integer) -> Bool
+    bit_literal(xs...)
 
-Return `true` if all masked position of index is 1.
+Create a [`BitStr`](@ref) by input bits `xs`.
 
 # Example
 
-`true` if all masked positions are 1.
-
 ```jldoctest
-julia> allone(bit"1011", 0b1011)
-true
-
-julia> allone(bit"1011", 0b1001)
-true
-
-julia> allone(bit"1011", 0b0100)
-false
-```
-"""
-allone(b::BitStr{T}, mask::Integer) where T = allone(b.val, T(mask))
-
-"""
-    ismatch(index::Integer, mask::Integer, target::Integer) -> Bool
-
-Return `true` if bits at positions masked by `mask` equal to `1` are equal to `target`.
-
-## Example
-
-```julia
-julia> n = 0b11001; mask = 0b10100; target = 0b10000;
-
-julia> ismatch(n, mask, target)
-true
+julia> bit_literal(1, 0, 1, 0, 1, 1)
+110101
 ```
 """
-ismatch(b::BitStr{T}, mask::Integer, target::Integer) where T = ismatch(b.val, T(mask), T(target))
-
-# NOTE: we have to return a vector here since this is calculated in runtime
-
-"""
-    baddrs(b::Integer) -> Vector
-
-get the locations of nonzeros bits, i.e. the inverse operation of bmask.
-"""
-baddrs(b::BitStr) = baddrs(b.val) |> bit(len=length(b))
-
-
-"""
-    bfloat(b::BitStr) -> Float64
-
-float view, with MSB 0 bit numbering.
-See also [wiki: bit numbering](https://en.wikipedia.org/wiki/Bit_numbering)
-"""
-bfloat(b::BitStr) = bfloat(b.val; nbits=length(b))
-
-"""
-    bfloat_r(b::Integer; nbits::Int) -> Float64
-
-float view, with bits read in inverse order.
-"""
-bfloat_r(b::BitStr) = bfloat_r(b.val; nbits=length(b))
-
-"""
-    bint(b; nbits=nothing) -> Int
-
-integer view, with LSB 0 bit numbering.
-See also [wiki: bit numbering](https://en.wikipedia.org/wiki/Bit_numbering)
-"""
-bint(b::BitStr) = b.val
-
-"""
-    bint_r(b; nbits::Int) -> Integer
+bit_literal(xs...) = bit_literal(xs)
+function bit_literal(xs::NTuple{N, T}) where {N, T<:Integer}
+    val = T(0)
+    for k in 1:N
+        xs[k] == 0 || xs[k] == 1 || error("expect 0 or 1, got $(xs[k])")
+        val += xs[k] << (k - 1)
+    end
+    return BitStr{N}(val)
+end
 
-integer read in inverse order.
-"""
-bint_r(b::BitStr) = breflect(b; nbits=length(b))
+basis(b::BitStr) = typemin(b):typemax(b)
```

Comment:
I think we need a `ZeroTo` iterator type which is similar to `Base.OneTo`. It will be faster.

Reply:
Not sure how much performance we can gain in practise. File an issue and do it later?

Reply:
Cool. Good to know.